### ISSUE TITLE
Multiple processes on a single GPU in Horovod? 

### ISSUE THREAD

**Issue by guoyuanxiong** [horovod/horovod#1886]
**Environment:**
1. Framework: (TensorFlow, Keras, PyTorch, MXNet)
2. Framework version:
3. Horovod version:
4. MPI version:
5. CUDA version:
6. NCCL version:
7. Python version:
8. OS and version:
9. GCC version:

**Checklist:**
1. Did you search issues to find if somebody asked this question before?
2. If your question is about hang, did you read [this doc](https://github.com/horovod/horovod/blob/master/docs/running.rst)?
3. If your question is about docker, did you read [this doc](https://github.com/horovod/horovod/blob/master/docs/docker.rst)?
4. Did you check if you question is answered in the [troubleshooting guide](https://github.com/horovod/horovod/blob/master/docs/troubleshooting.rst)?

**Your question:**
Please ask your question here.

I am trying to run a distributed learning algorithm with 8 workers on a 4-GPU machine. Is it possible to allocate two processes to a single GPU in Horovod?

**Comment by tgaddair** [comment]
Hey @guoyuanxiong, when using NCCL, this is not supported, as NCCL does not allow the same GPU to be assigned to multiple ranks.  

Can you give some more details about what it is you're trying to achieve?  There may be another way to do what you're trying to do.

**Comment by guoyuanxiong** [comment]
> Hey @guoyuanxiong, when using NCCL, this is not supported, as NCCL does not allow the same GPU to be assigned to multiple ranks.
> 
> Can you give some more details about what it is you're trying to achieve? There may be another way to do what you're trying to do.

**Comment by guoyuanxiong** [comment]
Thanks for your reply. I am trying to test the performance of my proposed distributed learning algorithm where multiple workers solve some optimization problems in parallel and exchange results iteratively. I want to simulate a non-trivial number of workers (e.g., 8 workers), but my machine only has four GPUs. My initial idea is to put several workers on a single GPU so that at least it works better than using only one GPU. Any idea on how shall I proceed given the current hardware restriction?

**Comment by coreyjadams** [comment]
I thought it might be useful to chime in here. I was trying to do the same thing, multiple ranks per GPU.  I hit the same issue.

Ultimately, I recompiled horovod and forced it to disable NCCL for communication.  The scaling overhead was marginally higher but I made up for it with a performance boost in throughput, even for the same problem size.

I recompiled horovod with this command:
`CC=mpicc CXX=mpicxx HOROVOD_CUDA_HOME=$CUDA_DIR HOROVOD_WITH_PYTORCH=1 HOROVOD_WITH_TENSORFLOW=1 HOROVOD_WITH_NCCL=0 python setup.py build`

(and then subsequent install)

This was on Summit, so power9 CPUs + V100 GPUs.  I was not using a traditional network but a sparse convolutional network.  The model bounces back and forth between CPU and GPU, doing hash maps on CPU and matmul on GPU.  As the batch size increases, the work on the CPU and the GPU is scaling at a different rate and eventually the CPU bottlenecks the throughput.  But the memory and compute usage on the GPU is leaving a lot on the table, unused.

Without MPS mode, and for a fixed network and batch size (192, peak throughput for 1 rank), this network achieved 5.9 Img/s.  Disabling NCCL and running more ranks on a single GPU with horovod, I saw increased throughput for the same problem:
2 ranks / GPU: 8.6 Img/s
4 ranks / GPU: 15.3 Img/s
8 ranks / GPU: 16.2 Img/s

All told, that's 2.7x speedup with exactly the same hardware.  It also scaled out over an entire summit node, 4 ranks per GPU, 6 GPUs per node.

So, @guoyuanxiong I think if you build horovod without NCCL you should be able to run your workload as you describe above.

**Comment by rllin** [comment]
@coreyjadams

i stumbled upon your experience here (you were right it was useful!)

I tried to reproduce this behavior in [colab](https://colab.research.google.com/drive/1pg20ZNqbV-zSWMZovNih-6w1buJwIBi3) but **could not**.

- i could not find the same `HOROVOD_WITH_NCCL` flag you used; i had to use `HOROVOD_WITH_MPI=1 HOROVOD_GPU_ALLREDUCE=MPI pip install -q horovod[tensorflow]`
- I used a single K80 that colab gives
- I instantiate a `tf.float32` of shape `[4096, 10_000]`
- I move this to device `GPU:0`
- I do a simple matmul (simple mlp)
- I measure the speed of this in `examples/s`

I run this with `2` vs `4` processes to see if I can get a speedup.

I get an average of `8.8e4 examples/s` per process for `2` and of `3.8e4 examples/s` for `4`.

Looking like total throughput stays fairly **constant.**

Do you have any insight?  Perhaps you managed to call your `horovodrun` correctly and I'm misunderstanding the CLI?  Thanks!

**Comment by coreyjadams** [comment]
Hi @rllin - I'm glad my notes served a purpose!  Looking at your notebook, I see that you don't enable MPS anywhere.  Did you do that?  I expect without the CUDA Multi-process Server running you will see no speedup, and in fact a slowdown as it has to switch contexts so much.
```bash
nvidia-cuda-mps-control -d
```

You can see the full MPS docs here: https://docs.nvidia.com/deploy/pdf/CUDA_Multi_Process_Service_Overview.pdf

Another thing to note: MPS really became good in the Volta generation of GPUs, and colab runs K80s.  So I don't think you'll see an improvement again because the cards don't have the right hardware support, those K80s are just ancient at this point.  Lately I'm running everywhere on Ampere and MPS makes a big difference.

Lastly, my app is using sparse convolutional networks.  So it has a big hashmap operation on CPU, and bounces over to GPU for a matmul.    Scaling up on the CPU is the bottle neck once the batch size gets big enough, and it becomes memory bound I think.  So breaking to smaller batches allows better overlap of CPU and GPU ops, and overall throughput improvement for me.

Since your test networks are small, you can likely get improvement with MPS but I am not sure, the CPU side is going to be low latency in your app.  Overloading like this really makes a big difference when you have CPU/GPU hybrid code.

**Comment by rllin** [comment]
@coreyjadams Thanks for such a thorough response.

I will absolutely give this a shot.

A bit more about my needs, I'm being bottlenecked by tensorflow's single cuda stream, hitting at best 8 GB/s on V100s which should have up to 16 GB/s.  I have very, very simple MLPs on the GPU but I have quite large bytes/example that I want to push from CPU to GPU every training step.  Ideally using multiple streams will let me actually saturate the hardware PCIE limit.

**Comment by rllin** [comment]
my skimming/google-fu is not strong enough.  if I have the mps server running, do i need to make `mpirun`/`horovodrun` aware of this server somehow?

**Comment by coreyjadams** [comment]
Hey @rllin here's my answers:

- your use case sounds like a good candidate for improved throughout. I bet it helps.
- no need to do anything to mpi and horovod except assigning the GPU correctly. The cuda multiprocess server does all the scheduling of kernels on the GPU from different cpu processes.

Please, let me know if it works and you see a speedup!

**Comment by rrrongon** [comment]
Hi I have similar kind of problem.I believe this thread is good place to discuss about the issue.
1. I am trying to setup a test bed like @guoyuanxiong . I have one machine with 1 GPU node. Initially, my plan was to test a distributed Deep learning solution with two process. Architecture would be data parallelism. So, two process having two model training and combine their gradiants. However, in my case i tried 
`horovodrun -np 2 -H localhost:2 python dataloader_test.py`
above code to spawn two processes. if I use horovod allreduce blocking function to reduce the gradients at the end of each batch iteration calculation, rank 0 blocks rank 1 to process the data and move forward. Where as Rank 0 waits for Rank 1 at allreduce point also not letting Rank 1 let in to process.

2. In case of Horovod, how it is different from Torch itself in terms of taking care of gradients and communication. Torch itself can take care of multiple process when we run code in multiple nodes. We don't take care of the params manually. So why do we need horovod for such case?

Could any one help me understand the situation?
